### PR TITLE
Concurrent CT write

### PR DESCRIPTION
**Pull request by ikatkov**
Moved branch ohadshacham:concurrentCTwrite to omid repo. No other changes.
Old PR - https://github.com/yahoo/omid/pull/75

Rebased, please comment below.
=====
caling up the TSO by parallelizing the batched writes to the commit table. The architecture mostly changed by adding batch events and creating handlers under PersistenceProcessorImpl.
This commit increases the performance of Omid from 55K tps to 380K tps when using 6 region servers.

Please note that the addition of MonitoringContext significantly degrades performance and therefore, should becomes optional to achieve good performance.

### CODE REVIEW

File: tso-server/src/main/java/org/apache/omid/tso/ReplyProcessorImpl.java [hunk 1]
```diff
@@ -27,106 +28,156 @@
...
 
-    final RingBuffer<ReplyEvent> replyRing;
+    final RingBuffer<ReplyBatchEvent> replyRing;
     final Meter abortMeter;
     final Meter commitMeter;
     final Meter timestampMeter;
 
+    public static final int NO_ORDER = (-1);
+
+    volatile private long nextIDToHandle;
```
Comment (ikatkov) [inline tso-server/src/main/java/org/apache/omid/tso/ReplyProcessorImpl.java]: long and volatile!
Needs to be either int or AtomicLong

File: tso-server/src/main/java/org/apache/omid/tso/ReplyProcessorImpl.java [hunk 2]
```diff
@@ -27,106 +28,156 @@
...
         replyExec.submit(replyProcessor);
 
         abortMeter = metrics.meter(name("tso", "aborts"));
         commitMeter = metrics.meter(name("tso", "commits"));
         timestampMeter = metrics.meter(name("tso", "timestampAllocation"));
+
+        nextIDToHandle = 0L;
+        futureEvents = new PriorityQueue<ReplyBatchEvent>(10, new Comparator<ReplyBatchEvent>() {
+            public int compare(ReplyBatchEvent replyBatchEvent1, ReplyBatchEvent replyBatchEvent2) {
+                return (Long.valueOf(replyBatchEvent1.getBatchID()).compareTo(replyBatchEvent2.getBatchID()));
```
Comment (ikatkov) [inline tso-server/src/main/java/org/apache/omid/tso/ReplyProcessorImpl.java]: return (Long.compare(replyBatchEvent1.getBatchID(), replyBatchEvent2.getBatchID()));

File: tso-server/src/main/java/org/apache/omid/tso/PersistenceProcessorImpl.java [hunk 3]
```diff
@@ -17,415 +17,154 @@
...
-        this.batch = batch;
-        this.batchPersistTimeoutInMs = config.getBatchPersistTimeoutInMs();
-        this.leaseManager = leaseManager;
-        this.commitTableClient = commitTable.getClient();
-        this.writer = commitTable.getWriter();
-        this.reply = reply;
-        this.retryProc = retryProc;
-        this.panicker = panicker;
+    volatile private long batchIDCnt;
+    volatile private long lowWatermark = INITIAL_LWM_VALUE;
```
Comment (francisco-perez-sorrosal) [inline tso-server/src/main/java/org/apache/omid/tso/PersistenceProcessorImpl.java]: long and volatile!
Needs to be either int or AtomicLong

**Comment by francisco-perez-sorrosal** [comment]
Squashed previous mess of commits into commit https://github.com/yahoo/omid/pull/89/commits/db84845872de501111a604224325d13611db8b4e

**Comment by francisco-perez-sorrosal** [comment]
Rebased on top of current master on commit f6220a8

**Pull request merged**